src/sys/unix/tcp.rs

```diff
@@ -202,6 +215,226 @@ pub(crate) fn get_send_buffer_size(socket: TcpSocket) -> io::Result<u32> {
     Ok(optval as u32)
 }
 
+pub(crate) fn set_keepalive(socket: TcpSocket, keepalive: bool) -> io::Result<()> {
+    let val: libc::c_int = if keepalive { 1 } else { 0 };
+    syscall!(setsockopt(
+        socket,
+        libc::SOL_SOCKET,
+        libc::SO_KEEPALIVE,
+        &val as *const _ as *const libc::c_void,
+        size_of::<libc::c_int>() as libc::socklen_t
+    ))
+    .map(|_| ())
+}
+
+pub(crate) fn get_keepalive(socket: TcpSocket) -> io::Result<bool> {
+    let mut optval: libc::c_int = 0;
+    let mut optlen = mem::size_of::<libc::c_int>() as libc::socklen_t;
+
+    syscall!(getsockopt(
+        socket,
+        libc::SOL_SOCKET,
+        libc::SO_KEEPALIVE,
+        &mut optval as *mut _ as *mut _,
+        &mut optlen,
+    ))?;
+
+    Ok(optval != 0)
+}
+
+pub(crate) fn set_keepalive_params(socket: TcpSocket, keepalive: TcpKeepalive) -> io::Result<()> {
+    use std::{fmt, error::Error};
+    #[derive(Debug)]
+    struct TcpKeepaliveError {
+        msg: &'static str,
+        inner: io::Error,
+    }
+
+    impl std::fmt::Display for TcpKeepaliveError {
+        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+            write!(f, "{}: {}", self.msg, self.inner)
+        }
+    }
+
+    impl Error for TcpKeepaliveError {
+        fn source(&self) -> Option<&(dyn Error + 'static)> {
+            Some(&self.inner)
+        }
+    }
+
+    fn error(msg: &'static str) -> impl Fn(io::Error) -> io::Error {
+        move |inner| io::Error::new(inner.kind(), TcpKeepaliveError { msg, inner })
```

Comment:
How much value does this additional message provide? Because `io::Error::new` allocates twice.

Reply:
Hmm, I thought it was potentially useful to know which keepalive parameter couldn't be set, but it may not be worth the extra allocation. I can remove this and just forward the OS error unmodified?

Alternatively, we could just make the `TcpKeepaliveError` type part of the public API and return it directly, avoiding the additional allocation from returning it behind an `io::Error`. However, that seemed inconsistent with the rest of the `TcpSocket` API.

Reply:
Let's just forward the error unmodified, seems like the easiest solution.